src/escalus_connection.erl

```diff
@@ -174,45 +186,121 @@ maybe_set_jid(Client = #client{props = Props}) ->
             throw({timeout, Name});
         {Stanza, _} ->
             Stanza
     end.
 
--spec get_stanza_with_metadata(client(), any(), timeout()) ->
-    {exml_stream:element(), metadata()}.
+-spec get_stanza_with_metadata(client(), any(), timeout()) -> {exml_stream:element(), metadata()}.
 get_stanza_with_metadata(Client, Name, Timeout) ->
-    case get_stanza_safe(Client, Timeout) of
+    get_stanza_with_metadata(Client, Name, Timeout, fun(_) -> true end).
+
+-spec get_stanza_with_metadata(client(), any(), timeout(), stanza_pred()) ->
+                                      {exml_stream:element(), metadata()}.
+get_stanza_with_metadata(Client, Name, Timeout, Pred) ->
+    case get_stanza_safe(Client, Timeout, Pred) of
         {error, timeout} ->
             throw({timeout, Name});
         StanzaWithMetadata ->
             StanzaWithMetadata
     end.
 
--spec get_stanza_safe(client(), timeout()) ->
-    {error, timeout} | {exml_stream:element(), map()}.
-get_stanza_safe(#client{rcv_pid = Pid, jid = Jid}, Timeout) ->
+-spec get_stanza_safe(client(), timeout()) -> {error, timeout} | {exml_stream:element(), metadata()}.
+get_stanza_safe(Client, Timeout) ->
+    get_stanza_safe(Client, Timeout, fun(_) -> true end).
+
+-spec get_stanza_safe(client(), timeout(), stanza_pred()) ->
+    {error, timeout} | {exml_stream:element(), metadata()}.
+get_stanza_safe(Client, Timeout, Pred) ->
+    Handler = fun(Stanza, Metadata) ->
+                      case Pred(Stanza) of
+                          true -> {finish, {Stanza, Metadata}};
+                          false ->
+                              handle_received_stanza(Client, Stanza, Metadata),
+                              continue
+                      end
+              end,
+    receive_stanzas(Client, Timeout, Handler).
+
+-spec wait(client(), timeout()) -> ok.
+wait(Client, Timeout) ->
+    Handler = fun(Stanza, Metadata) ->
+                      handle_received_stanza(Client, Stanza, Metadata),
+                      continue
+              end,
+    {error, timeout} = receive_stanzas(Client, Timeout, Handler),
+    ok.
+
+-spec wait_forever(client()) -> no_return().
+wait_forever(#client{event_client = EventClient, jid = Jid, rcv_pid = Pid} = Client) ->
     receive
         {stanza, Pid, Stanza, Metadata} ->
+            escalus_event:pop_incoming_stanza(EventClient, Stanza),
             escalus_ct:log_stanza(Jid, in, Stanza),
-            {Stanza, Metadata}
-    after Timeout ->
-              {error, timeout}
+            handle_received_stanza(Client, Stanza, Metadata),
+            wait_forever(Client)
+    end.
+
+receive_stanzas(Client, Timeout, Handler) ->
+    Tref = erlang:send_after(Timeout, self(), TimeoutMsg = {timeout, make_ref()}),
+    Result = do_receive_stanzas(Client, TimeoutMsg, Handler),
+    erlang:cancel_timer(Tref),
+    Result.
+
+do_receive_stanzas(#client{event_client = EventClient, jid = Jid, rcv_pid = Pid} = Client, TimeoutMsg, Handler) ->
```

Comment:
How about using `receive ... after TImeout -> {error, timeout} end `
instead of
`erlang:send_after(TImeout, self() , {error, timeout}) ... receive ...; TimeoutMessage -> {error, timeout} end ...` ??

It is very unlikely to happen, but from what I understand there is raise condition in this approach:
eg. if Timeout is 5000 ms and after 4995 ms process receives a message and logging takes 10 ms, than process is left with `Res` returned and `TimeoutMsg` hanging
Correct me if I'm wrong :)

Reply:
`receive .. after` cannot be used (there are multiple messages being received with a simple timeout, this is the whole point of this change) and the race condition has no effect as it would just result in a message in the mailbox that is never read. This implementation was used in escalus before, for a long time.

Reply:
@aleklisi after our discussion I have to note that the remark about the race condition is valuable. A cleanup added in line 267 would fix it: